storage/remote/codec.go

```diff
@@ -183,6 +183,148 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
 	}
 }
 
+func StreamChunkedReadResponses(
+	stream io.Writer,
+	queryIndex int64,
+	ss storage.SeriesSet,
+	sortedExternalLabels []prompb.Label,
+	sampleLimit int) error {
+	var numSamples = 0
+
+	for ss.Next() {
+		series := ss.At()
+		iter := series.Iterator()
+		lbls := MergeLabels(labelsToLabelsProto(series.Labels()), sortedExternalLabels)
+
+		for iter.Next() {
+			// TODO(bwplotka): Expose maxChunks as flag?
+			chks, samples, err := encodeChunks(iter, 120, 100, sampleLimit-numSamples)
+			if err != nil {
+				return err
+			}
+			numSamples += samples
+			if len(chks) == 0 {
+				break
+			}
+
+			b, err := proto.Marshal(&prompb.ChunkedReadResponse{
+				// TODO(bwplotka): Do we really need multiple?
+				ChunkedSeries: []*prompb.ChunkedSeries{
+					{
+						// TODO(bwplotka): Oneof? We don't need to send labels all the time.
+						Labels: lbls,
+						Chunks: chks,
+					},
+				},
+				QueryIndex: queryIndex,
+			})
+			if err != nil {
+				return errors.Wrap(err, "marshal ChunkedReadResponse")
+			}
+
+			if _, err := stream.Write(b); err != nil {
+				return errors.Wrap(err, "write to stream")
+			}
+		}
+
+		if err := iter.Err(); err != nil {
+			return err
+		}
+	}
+	if err := ss.Err(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// encodeChunks expects iterator to be ready to use (aka iter.Next() done before invoking).
+func encodeChunks(iter storage.SeriesIterator, maxSamplesPerChunk int, maxChunks int, sampleLimit int) ([]prompb.Chunk, int, error) {
+	var (
+		chkMint int64
+		chk     *chunkenc.XORChunk
+		app     chunkenc.Appender
+		err     error
+
+		isNext     = true
+		numSamples = 0
+		chks       = make([]prompb.Chunk, 0, maxChunks)
+	)
+
+	for isNext {
+		numSamples++
+		if sampleLimit > 0 && numSamples > sampleLimit {
```

Comment:
Is it still possible to hit this?

Reply:
Yes, why not? Sample limit is counted overall for all frames, so `sampleLimit` here contains the left samples we can serve. 

With streaming it does not makes sense much maybe as we made remote read quite efficient with this, but still caller would like to limit number of overall samples maybe in all frames together? What do you think?

Reply:
The below error is wrong though as `sampleLimit` is not overall limit, just the left samples for the chunks here. Will fix the error message.